### ISSUE TITLE
Clither Server Discussion

### ISSUE THREAD

**Issue by ThunderGemios10** [ClitherProject/Slither.io-Protocol#6]
So, We have some client and serverbound packets, I think its time to start developing Clither-Server
@Kogs Are you familiar with Java? If you are i would like to ask for help.
@Porama6400 Hopefully your not busy :)

**Comment by philw1508** [comment]
I am Not really a coder... just starten to. But my question is why you guys are interested in slither.io "hacks" ? I didnt See any Post of u in agario client or ogar feeder....

**Comment by ThunderGemios10** [comment]
@philw1508 We are not trying to hack Slither.io lol, We are making a server implementation (Like Ogar)

**Comment by Kogs** [comment]
I'm very familiar with Java. But currently i dont have the time to write the client.
To test the traffic so far I use a Node Js server. And so far its working very well.

**Comment by nightwolf93** [comment]
I have made a C# server, actually i have a BigEndian class ready, the Init packet with the username sent by the client implemented and the global score, i just need to get the full process of the game. I will put this on github tonight after the work if u wanna see how it's make.
If u need to contact for some collaboration like work on the server, write here :)

**Comment by ThunderGemios10** [comment]
@nightwolf93 We'd like to have you over! Although we are writing this in Java.

**Comment by nightwolf93** [comment]
I prefer to do it in .net, i actually hate Java haha. But we can work together, there is no problem :p

**Comment by ThunderGemios10** [comment]
@nightwolf93 Awww, Why does everyone hate Java? :/, Anyways, If you would like, You can rewrite your C# server and make a pull request and we'd be happy to invite you over :)

**Comment by nightwolf93** [comment]
I don't know,  probably because it's oracle, i am android developer at work, but im on Xamarin :p. I'm probably crazy haha. Why rewrite my C# server ?

**Comment by ThunderGemios10** [comment]
@nightwolf93 The fact that we are making Clither in Java :P

**Comment by nightwolf93** [comment]
Yeah, but i just want to continue on my server, but i can help for the protocol decryption. :)

**Comment by ThunderGemios10** [comment]
@nightwolf93 Aww, All right then :/ Any help is much appreciated!

**Comment by nightwolf93** [comment]
I will think about help you on the Java server, but i'm not sure, Java hate me and i hate Java :/

**Comment by ThunderGemios10** [comment]
@nightwolf93 Why does java hate you? and why do you hate java?

**Comment by nightwolf93** [comment]
I don't know haha. Im a Microsoft addict maybe.

**Comment by ThunderGemios10** [comment]
@nightwolf93 But if you know how to use C# then you know how to use Java too. (At least the import mechanic is similiar)

**Comment by nightwolf93** [comment]
I have no problem for writing Java code. I do some java at work, just it's not my best ;)

**Comment by ThunderGemios10** [comment]
@nightwolf93 That's fine, We just need any Java Developer, Beginner or Pro. :)

**Comment by nightwolf93** [comment]
I will think about that. This team have a slack or something like that ?

**Comment by ThunderGemios10** [comment]
@nightwolf93 Kogs is helping with the protocol, @Porama6400 is always busy = =, @SkorrloreGaming prefers to make his own games, And I, well i'm a person who likes to code APIs in Java.

**Comment by Porama6400** [comment]
No , i like API too :)

**Comment by ThunderGemios10** [comment]
@Porama6400 Your always busy, da :P

**Comment by Kogs** [comment]
@nightwolf93 @ThunderGemios10 

@circa94 and me started writing the server in javascript :D so we have 3 servers implemtations.
@nightwolf93 some help with the protocol would be very nice.

I have no Problem with C# and no Problem with Java. The Question is wich Language is the best for the server?

**Comment by circa94** [comment]
Kogs and me are currently developing a server with nodejs. It works good so far. We will update here the protocol if we find out some new stuff.

**Comment by ThunderGemios10** [comment]
@Kogs I think Java is best, Since it has multih-threading.. And it should improve performance, Node.js has one thread so if that thread fails the server's performance drops, So i don't want that, C#... well i'm not sure ...

**Comment by circa94** [comment]
I believe in my node server!

**Comment by Kogs** [comment]
Yes I think that with the node.js we cant handle up to 500 players with bots.

**Comment by ThunderGemios10** [comment]
@circa94 @Kogs It's been a struggle finding Java developers to help out :/ *cough* Ogar *cough*, But it seems everyone prefers Node.js over Java :/

**Comment by nightwolf93** [comment]
Pure Javascript or transcompiled language ?(like coffee)

**Comment by circa94** [comment]
@nightwolf93 pure javascript

**Comment by nightwolf93** [comment]
ok, for my part i prefer C#, but i can try to help you on your server, i can work at the same time on my own, btw my c# server is very crappy, it's just a poc :p

**Comment by ThunderGemios10** [comment]
@nightwolf93 We haven't even started yet. Clither right now is an empty shell full of errors :/

**Comment by circa94** [comment]
my server is a shell where login and fist initial steps are working. you can even see your username in the game

**Comment by nightwolf93** [comment]
I just have few packet implemented too.. :/ i have started this yesterday

**Comment by ThunderGemios10** [comment]
@circa94 @nightwolf93 Nice.. Looks like the two of your implementations has already progress... :/

**Comment by circa94** [comment]
about 500-600 lines of working code

**Comment by ThunderGemios10** [comment]
@circa94 That's lot :O

**Comment by nightwolf93** [comment]
@circa94 you have ur node js server on git on something else for looking how you handle the login ?

**Comment by Kogs** [comment]
500-600 lines are nothing.

**Comment by circa94** [comment]
i will discuss with Kogs if we maybe bring the server to github, so we all can help to develop it.

**Comment by nightwolf93** [comment]
@Kogs In javascript yeah xD

**Comment by Kogs** [comment]
@nightwolf93  we are coding in  http://cloud9.gg/

**Comment by nightwolf93** [comment]
@Kogs https://c9.io/ maybe ? xD

**Comment by Kogs** [comment]
yes :D

**Comment by circa94** [comment]
xDD

**Comment by ThunderGemios10** [comment]
@Kogs lol

**Comment by Kogs** [comment]
wtf xDD

**Comment by nightwolf93** [comment]
@Kogs Visual studio my friend .. :p

**Comment by Kogs** [comment]
I think for the final server cant use Javascript I think. The question is Java or C#?

**Comment by circa94** [comment]
sure u can...

**Comment by ThunderGemios10** [comment]
@Kogs *cheering repeatedly* Java! Java! Java!

**Comment by Kogs** [comment]
as far as I've seen the server has a ton of logic to be written (movement, collision) In Javascript its hard to manage that many logic. With a nice Java/C# Api it would be easier. I Think

**Comment by nightwolf93** [comment]
If i'm the only in C#, it's not worth :/ i can do some java, it's not a problem.

**Comment by Kogs** [comment]
I think Java is more protable (Linux server)

**Comment by circa94** [comment]
if you don't would waste your time to decide which language and just start coding...

**Comment by ThunderGemios10** [comment]
@Kogs I already written an API for clither https://github.com/ClitherProject/Clither-API

**Comment by Kogs** [comment]
I saw it yes. But is it final. Or just an first try?

**Comment by ThunderGemios10** [comment]
@circa94 Choosing a language is an important step to Programming...

**Comment by nightwolf93** [comment]
Let's go in Java, it's strong too ;)

**Comment by ThunderGemios10** [comment]
@Kogs Initial release, Might add more methods once Clither-Server is done

**Comment by Kogs** [comment]
@ThunderGemios10  kk

**Comment by circa94** [comment]
@nightwolf93  or C#? :D

**Comment by ThunderGemios10** [comment]
@circa94 Hmm... Never really used it... But is it fast?

**Comment by circa94** [comment]
@ThunderGemios10 is the 10 in your name your age?

**Comment by ThunderGemios10** [comment]
@circa94 Nope, Not anymore, This username was created when i was 10 years old why?

**Comment by Kogs** [comment]
This discussion is going nowhere at the moment...

**Comment by circa94** [comment]
okay that explains a lot, so i can not be angry when you ask silly questions. 

But to answer your question: yes c# is fast

**Comment by ThunderGemios10** [comment]
@Kogs Yes, Let's just vote.. I vote for Java!

**Comment by circa94** [comment]
Assembler!

**Comment by nightwolf93** [comment]
@circa94 of course

**Comment by ThunderGemios10** [comment]
@circa94 What?

**Comment by Kogs** [comment]
@ThunderGemios10 how good are your Java skills.

**Comment by ThunderGemios10** [comment]
@Kogs I'd say 2-4 xD, I'm still a basic Java Developer, I mostly love maintaining other projects rather than creating one.

**Comment by circa94** [comment]
@nightwolf93  and me coding in c# and @Kogs  and @ThunderGemios10  in Java

:D:D:D

**Comment by Kogs** [comment]
@circa94  that makes no sense at all :D

**Comment by ThunderGemios10** [comment]
@circa94 Your really trying to divide us don't ya?

**Comment by nightwolf93** [comment]
I have work on some big mmorpg server in c#, look on my github if u wanna see. So i can understand how to handle a lot of charge.

**Comment by circa94** [comment]
no. seriously. i give a shit which language. i want to run a fckn server

**Comment by ThunderGemios10** [comment]
@circa94 Le'ts use Java then? #JavaFTW

**Comment by circa94** [comment]
@nightwolf93 nice. i working a lot on revesing clash of clans network protokoll. also in c#

**Comment by nightwolf93** [comment]
@circa94 Yeah i saw this ;)

**Comment by ThunderGemios10** [comment]
@circa94 @nightwolf93 Ok, This discussion is starting to get off topic lol.

**Comment by Kogs** [comment]
I dont give a shit wich Language. Personaly i perefer Java but if 2 of 4 guys want to write it in c#, im fine with c#.

**Comment by nightwolf93** [comment]
I dont give a shit too let's go in java, simply the best way xD

**Comment by circa94** [comment]
then i'm out of here :)

**Comment by ThunderGemios10** [comment]
@Kogs @circa94 I sense a war coming...

**Comment by circa94** [comment]
no.. you know what? i will release my node server on github. and if anyone will join, he can. if not then not.

if i want to join your java project, i will do it..

**Comment by Kogs** [comment]
Guys I have some work to do now. If you decided wich langauge tell me.

**Comment by ThunderGemios10** [comment]
@Kogs I don't think @circa94 likes Java... I think he's allergic.. (metaphorically of course xD)

**Comment by circa94** [comment]
@ThunderGemios10 that's true

**Comment by ThunderGemios10** [comment]
***sigh*** This is Ogar, All over again, Node.js? or Java? .-.

**Comment by ThunderGemios10** [comment]
I bet if someone makes a Wings.io Server implementation, They will fight whether to use Node.js and Java.

**Comment by Kogs** [comment]
I think it should not be Node.js or Java. It should be Script Language or Object Orientated Language

**Comment by ThunderGemios10** [comment]
@Kogs I don't even know anymore.... :/ But i ain't giving up on Java! :P

**Comment by Kogs** [comment]
For now we will continue writing our node.js server. If we can play the game on it and the performance is not good we still can switch to c# or java. The hardest part of this Project is to get an idear of how the server should work and how the protocol is working.

**Comment by nightwolf93** [comment]
@Kogs Yes, i agree with that

**Comment by ThunderGemios10** [comment]
Right now, I'm using Ogar 2 (Java version) as reference.. or basically a template..

**Comment by rockthrower69** [comment]
Hi, I didn't read alle these comments.

I like the idea of making a Slither server. Don't care which language, but I do know performance is really important so you need to use a low-level programming language.

I used Ogar (NodeJS) before and noticed many unsolvable problems with the memory and CPU usage. NodeJS simply isnt suited for this with its 1, blocking thread.

Multithreading is a MUST in this project. I myself, can't code Java, but recommend it. Using NodeJS WILL turn out as a failure so it would be a waste of time making a server in Node.

Also, I hope you guys dont underestimate this. Slither as a game requires tons of math, way more than agario. I hope you are all prepared for it

Hope everything works ou :D.

**Comment by rockthrower69** [comment]
Also, C# seems a good language. Only disadvantage is its not multi-platform and you'll need Windows servers.

**Comment by ThunderGemios10** [comment]
@rockthrower69 Finally! Someone who understands!

**Comment by Kogs** [comment]
@rockthrower69 You can run C# in linux too. With mono.

**Comment by ThunderGemios10** [comment]
@Kogs Still, I think Java would be best, Since it has Multi-Threading.. Performance is always important! Plus Java is strong and powerful too!

**Comment by nightwolf93** [comment]
Just pick a language C# or Java. If we have more java developers let's go on Java is simply :)

**Comment by ThunderGemios10** [comment]
@nightwolf93 We'll use Java for Multi-Threading..

**Comment by rockthrower69** [comment]
Lets vote

**Comment by rockthrower69** [comment]
java

**Comment by rockthrower69** [comment]
C#

**Comment by Kogs** [comment]
I talked to @circa94 and we came to the decision that we will continue our node.js server. Not because its maybe not the best choice, but we never wrote an server Software for node.js. 

If you want/need some Help in the java server I can offer my services.

**Comment by nightwolf93** [comment]
@Kogs I can help on nodejs, im pretty god on it.

**Comment by circa94** [comment]
@nightwolf93 cool! we need some good input. i currently setup the git project.

**Comment by nightwolf93** [comment]
@circa94 Nice, tell me when it's done

**Comment by ThunderGemios10** [comment]
@Kogs All right then!, I'll ask for help when needed.

**Comment by ThunderGemios10** [comment]
**UPDATE: ** I pushed out some basic code to Clither-Server, As of now, It is broken and cannot be connected to via Slither.io. Will continue rolling out updates once more packets have been documented.

**Comment by ThunderGemios10** [comment]
@Kogs, Ok, seems i'm stuck now, How do you properly register Slither.io packets? Check Clither-Server for code, I am currently unable to connect, It seems there are missing packets or the packets aren't registered correctly..

**Comment by nightwolf93** [comment]
In your code i think u use a little endian buffer ? ( i have no check right now maybe u have modify this thing )

**Comment by ThunderGemios10** [comment]
@nightwolf93 What's wrong with it?

**Comment by nightwolf93** [comment]
Slither read in a big endian buffer

**Comment by Kogs** [comment]
@ThunderGemios10  Little tip for you. Try to use instance of as rarely. Its not very performant.

**Comment by ThunderGemios10** [comment]
@Kogs Noted. So I am registering the packets wrong or?

**Comment by nightwolf93** [comment]
This is just the direction when u read ur byte buffer or your write it. Take care of it, just use in packetdecoder or encoder a BigEndian byte order

**Comment by ThunderGemios10** [comment]
Ok, Can you give the BigEndian byte order?

**Comment by nightwolf93** [comment]
https://en.wikipedia.org/wiki/Endianness 👍

**Comment by ThunderGemios10** [comment]
@nightwolf93 I am using Big Endian, But still not connecting, I think the packets are not registered correctly..

**Comment by nightwolf93** [comment]
You have a timeout client-side ? or the client dont try to connect ?

**Comment by ThunderGemios10** [comment]
I think it is connected, But the world and snake is not loading..

**Comment by ThunderGemios10** [comment]
@nightwolf93 It repeatedly says "Client is connected" every second. So i'm assuming that it is not connecting..

**Comment by circa94** [comment]
@nightwolf93 repo is done. i think it can go on: https://github.com/circa94/JSlither

@allothers if you want, you can contribute to this repo also.

@ThunderGemios10 i will link your clither projects in my repo. it would be nice if you also do. also i will keep update the your network protocol repo in future and not create a new one.

**Comment by nightwolf93** [comment]
Here my repo : https://github.com/nightwolf93/SlitherNET

**Comment by nightwolf93** [comment]
Here my state for the server
Handshake - OK
Send the first snake (you) - OK

![image](https://cloud.githubusercontent.com/assets/2054618/14504873/de041838-01b6-11e6-9151-faf837ae52d9.png)

**Comment by Kogs** [comment]
We got it working.
@nightwolf93 some problems with the bytes. 

![2016-04-13 23_21_54-slither io](https://cloud.githubusercontent.com/assets/15261920/14509901/f228d50c-01ce-11e6-8c3c-492fd9508676.png)

**Comment by circa94** [comment]
@Kogs and me rule the world!

**Comment by nightwolf93** [comment]
@Kogs Haha pretty sure was that, can u update the file ? :p

**Comment by Kogs** [comment]
The Doc is right we just had Some wrong values. I'll push it tomorrow on our git repo (or maybe @circa94 now?)

**Comment by circa94** [comment]
i will push it later

**Comment by Porama6400** [comment]
@ThunderGemios10  do you know that you'll need to handshake with server first?

**Comment by nightwolf93** [comment]
@Kogs @circa94 Done, thank you guys !
![image](https://cloud.githubusercontent.com/assets/2054618/14512968/257dfcce-01e4-11e6-9721-6408f104ca40.png)

**Comment by ThunderGemios10** [comment]
@Porama6400 The handshaking function is suppose to work ? Unless its broken...

**Comment by Kogs** [comment]
@nightwolf93 next step is moving, food spawning and food collection

**Comment by nightwolf93** [comment]
@Kogs Yes, i will looking how to do it tonight.

**Comment by nightwolf93** [comment]
Food appear :)

![image](https://cloud.githubusercontent.com/assets/2054618/14539295/18ad3b94-0280-11e6-81f1-d8f9d535fceb.png)

**Comment by circa94** [comment]
I had it already done this morning

**Comment by nightwolf93** [comment]
Yeah i saw this right now !
The next step is how to move the snake, i have done some research with the G packet, but i dont know how to use it correctly right now

**Comment by Sentaq** [comment]
Nice guys! Just a question, is it gonna be a webclient?

**Comment by Kogs** [comment]
@Sentaq no or maybe :D currenty we just work an a server that means it runs only on your computer (or a server pc) and not in the web.

**Comment by Kogs** [comment]
@sentaq okay cool. But i think it will Take a while to the first stabel Release.

**Comment by Logan7432** [comment]
The main two things that need to be made is the client and the server. We need the client to run be able to run on web hosting sites, or be able to run on a home server for example appserv.

**Comment by ThunderGemios10** [comment]
@Logan7432 I do not recommend coding a client for Slither.io, Since you will have legal issues with the original devs.. Just code the server software...

**Comment by JosefKuchar** [comment]
It's legal to code raspberry pi client in c++?

**Comment by buzz** [comment]
Just for the record.
Overview of the official servers (from about 180 IPs):
https://goo.gl/CVrjqv

**Comment by SkorrloreGaming** [comment]
I am here to announce that I am now joining ClitherProject as a moderator.  Also does anyone know how to get net.packet.universal and unti packages to show up?

**Comment by ThunderGemios10** [comment]
Try asking @Kogs and @circa94 themselves.

**Comment by Kogs** [comment]
@SkorrloreGaming what do you mean with show up?

**Comment by crazyyi** [comment]
Wow this is a very interesting project. So is it coded in Java or C# now? I would like to know how do you guys make the client side look like the same as slither.io's. 

I thought this could be Node.js but after reading the discussion I felt dismal. Anyway I can do some java coding not sure if I could participate or not.

**Comment by Kogs** [comment]
@crazyyi We dont change the client.
Its only the Server impl. that we are doing.

**Comment by crazyyi** [comment]
@Kogs Yah just saw the usage so by injecting the sos command we can plug the server to the original client? Brilliant! But I have tried the JSlither and can't make it work with the original client side. :(

**Comment by sitano** [comment]
Hey guys, how the progress is going? What do you think / stuck with?

**Comment by Kogs** [comment]
The [JSlither](https://github.com/circa94/JSlither) project is on hold, because we dont have the time and no desire to finish it right now.

**Comment by RowanHarley** [comment]
Guys I'm working on a project (just wanted to mention it) [here](https://github.com/RowanHarley/Slither-Server/). I'm getting a lot of TypeError's though with invalid non-string/buffer chunks

**Comment by cyclingLinguist** [comment]
Hey guys, I'm trying to build a basic client and I have no idea how to calculate the snake's non-head pieces' movement. I'm trying to look through peoples homemade clients and servers to understand but I'm not making much progress. Any suggestions?

**Comment by RowanHarley** [comment]
Ok, I've got the sake to go in the right direction (yay!) but the movement's still buggy

**Comment by sitano** [comment]
I wrote server implementation in C++. Any body want to contribute?

**Comment by RowanHarley** [comment]
Sorry @sitano , I like js. If you know how to convert js to C++, then feel free to have a look around my server

**Comment by Mat2095** [comment]
@sitano : This sounds cool, in my opinion C++ is way better here than JS, I'll definitely look into it, though I'm not very good in C++ (too difficult ^^ ).
Make sure to add it [here](https://github.com/ClitherProject/Slither.io-Protocol/blob/master/CustomImplementations.md) or just leave a message here once it is published :)

**Comment by SkorrloreGaming** [comment]
Mat2095 don't try it, C++ is only supported on terminal. Also C++ is surprisingly slower then C and C#. I  would completely recommend you look into C# and not C++. Been coding on it for a long time now, and haven't seen any draw back or slowdowns with it. The speed is completely dependent on your hardware and efficiency of the code.

**Comment by Mat2095** [comment]
@SkorrloreGaming I do have some experiences with C and C++, I'd say around 200 hours (not including time to compile ^^). Also, why would I want anything else than terminal for a server??
C# is slower than C++ in most cases, sorry ;)
I personally prefer Java (been coding with it for ~8 years now), but this is more because of my personal taste.

**Comment by ThunderGemios10** [comment]
@Mat2095 Finally someone who understands :O

**Comment by SkorrloreGaming** [comment]
lol, hey Gemios what if I told you I am developing a hard link between minecraft and softly irc. making it interlace-able on such a link that you can get full control of the machine at hand. Check it out in my repositories 'HardScene_Bukkit'

**Comment by ThunderGemios10** [comment]
@SkorrloreGaming Yeah, I already checked it out. It looks awesome so far.
Also, Please don't call me "Gemios". Just thunder :P

**Comment by Speentie8081** [comment]
Hello World, I am aiming to get everyone to sue MineChat for using the minecraft API without permission. As well as adverting HardScene for Bukkit as a replacement to allow a Softly IRC client to connect to the minecraft server cleverly leveraging an extra port and thread to emulate connecting to the minecraft server completely legally in the form of a Bukkit plugin.

**Comment by CreaModZ** [comment]
Hello'! 

Some news about Slither server implementation?

**Comment by ThunderGemios10** [comment]
@CreaModZ Unfortunately no. Development on the server implementation has ceased due to lack of active developers.
